### PR TITLE
deprecate passing ModelCheckpoint instance to Trainer(checkpoint_callback=...)

### LINKED ISSUE #4014: resume_from_checkpoint loads duplicate ModelCheckpoint
## 🐛 Bug

When reloading Trainer from a checkpoint, the model callback appears twice in the callbacks list.
This ONLY happens if one provides both a checkpoint_callback AND custom callbacks list.

### To Reproduce

Code below 

#### Code sample

```python 
import torch
from torch.utils.data import Dataset
from pytorch_lightning import Trainer, LightningModule, Callback
from pytorch_lightning.callbacks import ModelCheckpoint


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    def training_epoch_end(self, outputs) -> None:
        torch.stack([x["loss"] for x in outputs]).mean()

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]


def run_test():
    # fake data
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    class RandomCallback(Callback):
        pass

    model = BoringModel()
    checkpoint_callback = ModelCheckpoint()
    trainer_args = dict(
        max_epochs=2,
        logger=False,
        progress_bar_refresh_rate=0,
        checkpoint_callback=checkpoint_callback,
        weights_summary=None,
        callbacks=[RandomCallback()]  # try to remove it and see what happens :)
    )
    trainer = Trainer(**trainer_args)
    trainer.fit(model, train_dataloader=train_data)
    print(trainer.callbacks)
    trainer = Trainer(**trainer_args, resume_from_checkpoint=trainer.checkpoint_callback.best_model_path)
    trainer.fit(model, train_dataloader=train_data)
    print(trainer.callbacks)


if __name__ == '__main__':
    run_test()

```

Produces:

```python 
callbacks = [ModelCheckpoint]   # first training
callbacks = [ModelCheckpoint, ModelCheckpoint]  # after resuming
```

### Expected behavior

```python 
callbacks = [ModelCheckpoint]   # first training
callbacks = [ModelCheckpoint]  # after resuming
```

### Environment
master branch 0.10.0
python 3.7
torch 1.5.1

### LINKED ISSUE #4386: After trainer resume_from_checkpoint checkpoint_callback doesn't restore best metric value
After I restore training with 
```
    pl_model = LightningModel.load_from_checkpoint(str(ckpt_path))

    trainer = Trainer(
        resume_from_checkpoint=str(ckpt_path.name),
        logger=instantiate(cfg.logger, experiment_key=cfg.experiment_id),
        checkpoint_callback=instantiate(cfg.checkpoint.model_checkpoint),
        callbacks=[instantiate(callback) for callback in cfg.callbacks],
    )
    trainer.fit(pl_model)
```
instantiate from facebook/hydra
The training process doesnt restore the best metric value and starts from scratch so I'm getting on the first restored epoch
```
INFO:lightning:Epoch 1129: val_mae reached 0.10418 (best 0.10418), saving model to
```
When there was a better checkpoint before interruption

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Pitch: Deprecate passing `ModelCheckpoint` instance directly to `checkpoint_callbacks` argument. Allow only bool. 

### FAQ

**Q:** Why can we not just remove the `checkpoint_callback` argument all together?
**A:** Lightning philosophy is to provide good defaults, this includes checkpointing. We need a way to turn it off though, i.e., `checkpoint_callback=False`

**Q:** Why can't we keep it the way it is?
**A:** When we add `Trainer(checkpoint_callback=mycustomcallback)`, internally Trainer adds it to `self.callbacks`. 
A problem arises when we reload the Trainer using `resume_from_checkpoint="..."`. In this case, we would end up with a conflict and a decision needs to be made: Do we load the callback that was saved in the checkpoint, or do we overwrite it with the one that is passed in to Trainer. We can try to handle this case and make the decision for the user (#4027) or we go with this PR here to remove the ambiguity alltogether.

### Alternatives
- Drop `checkpoint_callback`. This implies no default checkpointing. 
- Suggestion by @jeremyjordan: Remove `checkpoint_callback` and if `trainer.callbacks` is empty, add a ModelCheckpoint automatically. If any other callbacks need to be passed in, we need to supply a ModelCheckpoint manually.
- Keep as is, go with #4027 for the fix, document well the behaviour when resuming from checkpoint.

### TODOs for this PR:

- [ ] update docs -> follow up PR

Fixes #4014 automatically
Maybe Fixes #4386 (need to reproduce first)
Closes #3990 
Closes #4027
Related to #4335, as it is a good step in to the direction of allowing multiple model checkpoints.

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/callback_connector.py:71]: This is a bit confusing, the message makes me think that doing:

```python
Trainer(callbacks=[ModelCheckpoint])
```

is not supported, but that is exactly what this PR intends to support.

Did you mean?

```suggestion
            raise MisconfigurationException(
                "Trainer was configured with checkpoint_callback=False but found ModelCheckpoint"
                " in the trainer.checkpoint_callbacks list."
            )
```

File: pytorch_lightning/trainer/trainer.py
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:88]: > Q: Why can we not just remove the checkpoint_callback argument all together?
> A: Lightning philosophy is to provide good defaults, this includes checkpointing. We need a way to turn it off though, i.e., checkpoint_callback=False

What about keeping the `checkpoint_callback` parameter but changing its default value to `False`? I think it will be pretty annoying to have to set `checkpoint_callback=False` every time you pass a custom `ModelCheckpoint` via callbacks. And I think most people use a custom `ModelCheckpoint` instead of just `checkpoint_callback=True`

File: tests/models/test_restore.py
Comment (carmocca) [inline tests/models/test_restore.py:63]: I think this test is quite poor. Can you expand it to check that the `Trainer(resume_from_checkpoint=...)` checkpoint was restored correctly? Also that there are no extra ModelCheckpoints to avoid a regression of #4027

File: pytorch_lightning/trainer/trainer.py
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:88]: > What about keeping the checkpoint_callback parameter but changing its default value to False? 

doesn't solve the problem I'm trying to solve here, which is eliminate ambiguity when restoring the state of trainer. see answer of 2nd FAQ question.

> pretty annoying to have to set checkpoint_callback=False every time you pass a custom ModelCheckpoint

with this PR proposal, the value will be ignored if you pass in a custom one. False is only needed when you want to disable checkpointing completely. I believe I have this covered in a test.

File: tests/models/test_restore.py
Comment (awaelchli) [reply tests/models/test_restore.py:63]: yes of course, I will do that. thanks for the suggestion. 
I'm currently waiting for a decision by core about the api change. we are currently discussing this.
if this is accepted, I will finish this PR properly and we can deal with the details.

File: pytorch_lightning/trainer/connectors/callback_connector.py
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/callback_connector.py:71]: the message says that `Trainer(checkpoint_callback=False, callbacks=[ModelCheckpoint()])` conflicting. Either do 
`Trainer(checkpoint_callback=False, callbacks=[])` or
`Trainer(checkpoint_callback=True (default) , callbacks=[ModelCheckpoint()])`

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/callback_connector.py:71]: Okay, I thought it was the other way around. :+1:

File: tests/models/test_restore.py
Comment (carmocca) [reply tests/models/test_restore.py:63]: Thanks a lot for your time!

Regardless of the decision, I think #4027 is a high priority bug to fix. I don't know what alternatives are being discussed but this one seems sound :rocket:

Comment (awaelchli) [reply tests/models/test_restore.py:63]: True, I agree. I definitely want to get this done this week.

**Comment by edenlightning** [comment]
@williamFalcon @teddykoker remove ambiguity by making checkpoint_callback boolean. sg?

File: pytorch_lightning/tuner/batch_size_scaling.py
Comment (awaelchli) [inline pytorch_lightning/tuner/batch_size_scaling.py:147]: I removed these from Tuner because ModelCheckpoint now lives in callbacks list, and this is properly backed up by Tuner already.

**Comment by awaelchli** [comment]
I propose to update docs in a follow up PR, otherwise it will a bit too much to review at once.

File: pytorch_lightning/tuner/batch_size_scaling.py
Comment (awaelchli) [reply pytorch_lightning/tuner/batch_size_scaling.py:147]: cc @SkafteNicki

File: tests/models/test_restore.py
Comment (awaelchli) [reply tests/models/test_restore.py:64]: @carmocca so core team agrees with the changes. feel free to give me a tough review :)) 
I decided not to touch this test here too much because it was for a bugfix and only aims at testing the model properties. 
Instead, I added a new test that asserts the callbacks state is reloaded properly.

File: pytorch_lightning/trainer/trainer.py
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:88]: i agree with @carmocca , this is super confusing when adding my own checkpoint callback. given how loose the default checkpoint callback is, and with the coming customizations, I'd rather drop the `checkpoint_callback` arg altogether and force everything to be configured through `callbacks`. Given the implementation already exists, I personally don't think it's much of a request for people to create a callback (and confirm their settings while doing so) and pass it along

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:88]: yes fine with me, I don't have strong preference here. 
Note that this PR does NOT close the disussion on whether there should be a checkpoint_callback arg or not. 
I'm simply restricting what can be passed to the argument.

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/properties.py
Comment (carmocca) [inline pytorch_lightning/trainer/properties.py:198]: Is this required or just for convenience?

Thinking as a noob user. If I saw `trainer.checkpoint_callback` I would think it's the only one available, when in reality it's the first one.

If possible, I suggest renaming it to `first_checkpoint_callback` or, even better, removing it. If we want to support multiple callbacks, why have a property for the first one when order is not supposed to matter?

Comment (awaelchli) [reply pytorch_lightning/trainer/properties.py:198]: yes, good observation. 
I made this mainly for compatibility with a few places where Trainer needs to access the checkpoint callback. One example is the test method which can load the best model, and this is done through the  best model path saved in the checkpoint callback. Right now, this is for convenience and backward compatibility. 

I propose to keep this at least until deprecation of of ModelCheckpoint through checkpoint_callback argument expires.
In the meantime, we can find a way to handle multiple ckpt callbacks all around the trainer in a consistent way.

File: tests/models/test_restore.py
Comment (carmocca) [inline tests/models/test_restore.py:117]: LGTM. Just to be extra safe, mind adding the following test? So references are explicitly checked:


```suggestion

def test_callbacks_references_resume_from_checkpoint(tmpdir):
    """ Test that resuming from a checkpoint sets references as expected. """
    model = EvalModelTemplate()
    args = {'default_root_dir': tmpdir, 'max_steps': 1, 'logger': False}

    # initial training
    checkpoint = ModelCheckpoint(dirpath=tmpdir, monitor="early_stop_on", save_last=True)
    trainer = Trainer(**args, callbacks=[checkpoint])
    assert checkpoint is trainer.callbacks[0] is trainer.checkpoint_callback
    trainer.fit(model)

    # resumed training
    new_checkpoint = ModelCheckpoint(dirpath=tmpdir, monitor="early_stop_on", save_last=True)
    # pass in a new checkpoint object, which should take
    # precedence over the one in the last.ckpt file
    trainer = Trainer(**args, callbacks=[new_checkpoint], resume_from_checkpoint=str(tmpdir / "last.ckpt"))
    assert checkpoint is not new_checkpoint
    assert new_checkpoint is trainer.callbacks[0] is trainer.checkpoint_callback
    trainer.fit(model)


```

**Review by carmocca — APPROVED** [review]

File: tests/models/test_restore.py
Comment (awaelchli) [reply tests/models/test_restore.py:117]: Yeah this is a great test because it fails on master with the bug we fix here. 
I was actually first confused why it failed when I didn't realize I was on master :)

File: pytorch_lightning/trainer/trainer.py
Comment (ydcjeff) [inline pytorch_lightning/trainer/trainer.py:174]: **If ``True``, will configure a default ModelCheckpoint callback.**
So to pass custom ModelCheckpoint, it needs to be false?
But, answer is no, right?
Just to be sure for this case (appeared in mind when I first read it)
```suggestion            
            checkpoint_callback: If ``True``, enable checkpointing.
                It will configure a default ModelCheckpoint callback
                if there is no custom ModelCheckpoint callback in `:paramref:~Trainer.callbacks`.
                Default: ``True``.

                .. warning:: Passing a ModelCheckpoint instance to this argument is deprecated since
                    v1.1.0 and will be unsupported from v1.4.0.
```

**Review by justusschock — APPROVED** [review]

**Comment by awaelchli** [comment]
@s-rog Yes correct. I have already updated these docs you mention on another branch, which will follow up to this. I didn't want to make the PR too big, so keeping it separate.

**Review by s-rog — APPROVED** [review]

**Review by ydcjeff — APPROVED** [review]

**Pull request merged**
